Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```diff
@@ -90,62 +93,64 @@
     if (requestedAllHostsPattern && !m_requestedOptionalAccessToAllHosts)
         m_requestedOptionalAccessToAllHosts = YES;
 
-    __block MatchPatternSet grantedPatterns;
-    __block PermissionsSet grantedPermissions;
-    auto delegate = extensionController()->delegate();
+    class ResultHolder : public RefCounted<ResultHolder> {
+    public:
+        static Ref<ResultHolder> create() { return adoptRef(*new ResultHolder()); }
+        ResultHolder() = default;
+
+        bool matchPatternsAreGranted { false };
+        bool permissionsAreGranted { false };
+    };
+
+    Ref resultHolder = ResultHolder::create();
+
+    // Empty requests are automatically granted and don't call the delegate.
+    resultHolder->matchPatternsAreGranted = matchPatterns.isEmpty();
+    resultHolder->permissionsAreGranted = permissions.isEmpty();
 
-    auto originsCompletionHandler = ^(NSSet<_WKWebExtensionMatchPattern *> *allowedOrigins) {
-        // No granted origins were sent back, but origins were requested.
-        if (!allowedOrigins.count && origins.size()) {
+    Ref callbackAggregator = CallbackAggregator::create([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), resultHolder, permissions, matchPatterns]() mutable {
+        if (!resultHolder->matchPatternsAreGranted || !resultHolder->permissionsAreGranted) {
```

Comment:
Should this be an `||` or an `&&`? Or does doing this wait until they are both set?

Reply:
Or does `completionHandler(false)` just mean the callback aggregator isn't done yet?

Reply:
This is the response to the web process, and the `permissions.request()` requires all permissions to be granted (`true`) or none (`false`).